[ALLUXIO-3184] Master communication for server configuration checker

LuQQiu opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-3184

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19465/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19466/<h2>Build result: FAILURE</span></h2>[...truncated 2372 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19469/
Test PASSed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -64,6 +86,23 @@
 public class AlluxioMasterProcess implements MasterProcess {
   private static final Logger LOG = LoggerFactory.getLogger(AlluxioMasterProcess.class);
 
+  // Master metadata management.
```
With all this new functionality, I think it might be time to split the `MetaMaster` off into its own master implementation, similar to `FileSystemMaster` or `BlockMaster`. Otherwise this class is getting very long and complex.

LuQQiu replied:
Agree! To make AlluxioMasterProcess simple and have a separate MetaMaster. Could I have a separate PR to split AlluxioMasterProcess and MetaMaster? I think this part @apc999 @gpang may have different ideas.

gpang replied:
Yeah, having a "true" metamaster would be good.

aaudiber replied:
Separate PR is fine

aaudiber commented on core/server/master/src/main/java/alluxio/master/MasterProcess.java:
```diff
@@ -92,4 +94,29 @@ private Factory() {} // prevent instantiation
+   *
+   * @param masterId the master id of the standby master registering
+   * @param options the options that contains master configuration
+   * @throws NoMasterException if masterId cannot be found
```
why does `masterHeartbeat` use `boolean` to indicate that the master ID wasn't found, but `masterRegister` uses `NoMasterException`?

LuQQiu replied:
In the metaMasterSync which manages the master communications, I use the logic that if master id is not set, the masterHeartbeat will ask the master to re-register. When we register, we always get the master Id first and then call masterRegister. If everything goes as expected, the masterId should be found and will not trigger the NoMasterException.
```
MetaMasterSync.heartbeat()
      boolean shouldReRegister = mMasterClient.heartbeat(mMasterId.get());
      if (shouldReRegister) {
        mMasterId.set(mMasterClient.getId(mMasterHostname));
        registerWithMaster();
      }
```

So that masterId is not found is reasonable and expected in masterHeartbeat and not reasonable and expected in masterRegister.

aaudiber replied:
Got it, using exception for one but not the other SGTM

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:
```diff
@@ -0,0 +1,122 @@
+ * This class is not thread safe, so external locking is required.
+ */
+@NotThreadSafe
+public final class MetaMasterInfo {
```
Since this is info about a master, could we call it just `MasterInfo`?

LuQQiu replied:
better name! Changed!

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  /**
+   * Marks the master as registered, while updating all of its metadata.
```
what metadata?

LuQQiu replied:
Remove isRegister, it is not needed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  /**
+   * Updates the last updated time of the worker in ms.
```
worker -> master

LuQQiu replied:
Done, thanks!

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  /**
+   * @return the id of the wmaster
```
master

LuQQiu replied:
Done, thanks!

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:
```diff
@@ -0,0 +1,122 @@
+  public MetaMasterInfo(long id, String hostname) {
+    mHostname = Preconditions.checkNotNull(hostname, "hostname");
+    mId = id;
+    mStartTimeMs = System.currentTimeMillis();
```
This is more like the time the master registered. With failover the other master could have started much earlier than this time.

LuQQiu replied:
Yes! It will not show the actual start time.

MetaMasterInfo now only contains three values: ``` hostname, id, lastUpdatedTimeMs```. Those are the values that we need now. If other values are needed, we could add them later.

Thanks for pointing out that.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterMasterClient.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * The method the standby master should periodically execute
```
Sends a heartbeat to the master.

Standby masters periodically execute this method so that the primary master knows they are still running.

LuQQiu replied:
Changed, thanks!

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterMasterClient.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * The method the standby master should execute to register with the leader master.
```
Registers with the leader master.

LuQQiu replied:
Changed, thanks!

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterMasterServiceHandler.java:
```diff
@@ -0,0 +1,107 @@
+      @Override
+      public String toString() {
+        return String
+            .format("getMasterId:masterHostname=%s, options=%s", masterHostname, options);
```
nit: space after `:`

LuQQiu replied:
Changed, thanks

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,113 @@
+ * then sends it to the leader master. After which, the task will wait for the elapsed time
+ * since its last heartbeat has reached the heartbeat interval. Then the cycle will continue.
+ *
+ * If the task fails to heartbeat to the leader master, it will destroy its old master client
```
is this necessary? I think the `MetaMasterClient` should be able to handle the retries.

LuQQiu replied:
I think this line can be deleted, but it is the executor service that handle the retries.

```
  private final ExecutorService mExecutorService;
  protected void startMasters(boolean isLeader) {
    try {
      mWorkerConfigChecker.reset();
      if (isLeader) {
        mSafeModeManager.notifyPrimaryMasterStarted();

        //  The service that detects lost standby master nodes
        mExecutorService.submit(new HeartbeatThread(
            HeartbeatContext.MASTER_LOST_MASTER_DETECTION,
            new LostMasterDetectionHeartbeatExecutor(),
            (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      } else {
        // Standby master should setup MetaMasterSync to communicate with the leader master
        mMetaMasterSync = new MetaMasterSync(mMasterId, mMasterHostname, mMetaMasterClient);
        mExecutorService.submit(new HeartbeatThread(HeartbeatContext.MASTER_SYNC, mMetaMasterSync,
                (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      }
```

aaudiber replied:
The heartbeat will eventually provide retries since it keeps running periodically, but I think the client itself also provides retries: https://github.com/Alluxio/alluxio/pull/7233/files#diff-8495e238f9f1a6a00192d61f41daddd9R89

LuQQiu replied:
I see. Both the executor service and the client itself provide retries. I will delete this line.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,113 @@
+ * Task that carries out the necessary standby master to leader master communications, including
+ * register and heartbeat. This class manages its own {@link MetaMasterMasterClient}.
+ *
+ * When running, this task first requests master configuration from the standby master,
```
this sounds weird since the task is a part of the standby master, so the master would be requesting configuration from itself. Can we re-phrase?

LuQQiu replied:
Re-phrased, PTAL at the new one. Thanks

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,113 @@
+   */
+  private final AtomicReference<Long> mMasterId;
+
+  /** The hostname of the master. */
```
specify where this is the primary master hostname or this master's hostname.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,113 @@
+  @Override
+  public void heartbeat() {
+    try {
+      boolean shouldReRegister = mMasterClient.heartbeat(mMasterId.get());
```
Can the heartbeat return a `Command`, similar to the worker heartbeat? There might be new commands in the future, and Command.REGISTER is easier to understand than `true`

LuQQiu replied:
Should we make use of the same Command as what block master and file system master use now?
```
enum CommandType {
  Unknown = 0,
  Nothing = 1,
  Register = 2, // Ask the worker to re-register.
  Free = 3,     // Ask the worker to free files.
  Delete = 4,   // Ask the worker to delete files.
  Persist = 5,  // Ask the worker to persist a file for lineage
}
struct Command {
  1: CommandType commandType
  2: list<i64> data
}
```
This structure may be too complicated for meta master. I will prefer to use another Command.
```
enum MetaCommand {
  Unknown  = 0,
  Nothing = 1,
  Register = 2,
}
```

aaudiber replied:
I agree, `MetaCommand` is much better

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19473/<h2>Build result: ABORTED</span></h2>[...truncated 4405 lines...][INFO] Alluxio Integration ................................ SUCCESS [  1.703 s][INFO] Alluxio Integration - Mesos ........................ SUCCESS [  7.559 s][INFO] Alluxio Assembly - Server .......................... SUCCESS [  5.879 s][INFO] Alluxio Integration Checker ........................ SUCCESS [ 29.326 s][INFO] Alluxio Core - Client - HDFS ....................... SUCCESS [ 13.164 s][INFO] Alluxio Core - Client - Runtime .................... SUCCESS [ 11.444 s][INFO] Alluxio Integration - FUSE ......................... SUCCESS [ 11.223 s][INFO] Alluxio MiniCluster ................................ SUCCESS [12:14 min][INFO] Alluxio Tests ...................................... FAILURE [  01:39 h][INFO] Alluxio Under File System - GCS .................... SKIPPED[INFO] Alluxio Under File System - HDFS ................... SKIPPED[INFO] Alluxio Under File System - Aliyun OSS ............. SKIPPED[INFO] Alluxio Under File System - S3A .................... SKIPPED[INFO] Alluxio Under File System - Swift .................. SKIPPED[INFO] Alluxio Under File System - Microsoft Azure Blob Storage SKIPPED[INFO] Alluxio Integration - YARN ......................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 01:59 h[INFO] Finished at: 2018-05-04T13:26:26-07:00Archiving artifacts[INFO] Final Memory: 149M/2975M[INFO] ------------------------------------------------------------------------
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19474/<h2>Build result: ABORTED</span></h2>[...truncated 4400 lines...][INFO] Alluxio Integration ................................ SUCCESS [  1.446 s][INFO] Alluxio Integration - Mesos ........................ SUCCESS [  7.239 s][INFO] Alluxio Assembly - Server .......................... SUCCESS [  5.574 s][INFO] Alluxio Integration Checker ........................ SUCCESS [ 28.337 s][INFO] Alluxio Core - Client - HDFS ....................... SUCCESS [ 12.428 s][INFO] Alluxio Core - Client - Runtime .................... SUCCESS [ 11.201 s][INFO] Alluxio Integration - FUSE ......................... SUCCESS [ 10.565 s][INFO] Alluxio MiniCluster ................................ SUCCESS [12:14 min][INFO] Alluxio Tests ...................................... FAILURE [  01:39 h][INFO] Alluxio Under File System - GCS .................... SKIPPED[INFO] Alluxio Under File System - HDFS ................... SKIPPED[INFO] Alluxio Under File System - Aliyun OSS ............. SKIPPED[INFO] Alluxio Under File System - S3A .................... SKIPPED[INFO] Alluxio Under File System - Swift .................. SKIPPED[INFO] Alluxio Under File System - Microsoft Azure Blob Storage SKIPPED[INFO] Alluxio Integration - YARN ......................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 01:59 h[INFO] Finished at: 2018-05-04T13:46:29-07:00Archiving artifacts[INFO] Final Memory: 142M/2443M[INFO] ------------------------------------------------------------------------
Test FAILed.

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -232,6 +305,74 @@ public boolean isServing() {
+    if (master == null) {
+      throw new NoMasterException(ExceptionMessage.NO_MASTER_FOUND.getMessage(masterId));
+    }
+    synchronized (master) {
```
what is this `synchronized` for?

LuQQiu replied:
Delete the synchronized, thanks!
We do not need synchronized here for justing updating one MetamasterInfo.

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -232,6 +305,74 @@ public boolean isServing() {
+      master.updateLastUpdatedTimeMs();
+    }
+
+    List<ConfigProperty> configList = options.getConfigList().stream()
```
Will an option always have a configList?

LuQQiu replied:
For master communication, we do not need to support backward compatibility. If the standby masters are in old versions, it will not register with the leader master. If it is in the new version, it will always register with the leader master with configuration.

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -272,6 +413,17 @@ protected void startMasters(boolean isLeader) {
         mSafeModeManager.notifyPrimaryMasterStarted();
+
+        //  The service that detects lost standby master nodes
+        mExecutorService.submit(new HeartbeatThread(
```
if this becomes a standby, where does this executor service get shutdown or stopped?

LuQQiu replied:
Add stopExecutorService in startMasters. When masters switch position, it will first shut down its executor service and then create a new executor service.
```
  protected void startMasters(boolean isLeader) {
    try {
      mWorkerConfigChecker.reset();
      if (isLeader) {
        stopExecutorService();
        mExecutorService = mExecutorServiceFactory.create();
        mExecutorService.submit....
      } else {
        stopExecutorService();
        mExecutorService = mExecutorServiceFactory.create();
        mExecutorService.submit...
    } ...
```

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -272,6 +413,17 @@ protected void startMasters(boolean isLeader) {
+      } else {
+        // Standby master should setup MetaMasterSync to communicate with the leader master
+        mMetaMasterSync = new MetaMasterSync(mMasterId, mMasterHostname, mMetaMasterClient);
+        mExecutorService.submit(new HeartbeatThread(HeartbeatContext.MASTER_SYNC, mMetaMasterSync,
```
When becoming a primary, does this have to be stopped?

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -457,4 +643,26 @@ private void workerLostHandler(long id) {
+   * Sets the master id and hostname.
+   */
+  private void setMasterIdAndHostname() {
+    mMasterHostname = Configuration.get(PropertyKey.MASTER_HOSTNAME);
```
With fault tolerant masters, will this property always be set to the machines hostname? I feel like this property may not always be the machine's hostname.

LuQQiu replied:
The masterHostname here only be used for the leader master to identify a node and assign an id. I think if the value of MASTER_HOSTNAME is unchanged at runtime and no two masters have the same value, using MASTER_HOSTNAME as a representative of this node should be fine.

gpang replied:
Well, I don't think masters use `MASTER_HOSTNAME`. (this is a confusing part of the config) I think `MASTER_HOSTNAME` is only used for clients accessing the master to connect to the master.

I think this might be the first time the master uses the `MASTER_HOSTNAME` for its own hostname, so I don't think this will always be set correctly. Instead, maybe we should use: https://github.com/Alluxio/alluxio/blob/v1.7.1/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java#L277

aaudiber replied:
In HA mode masters need their own `alluxio.master.hostname` so that they can register themselves with zookeeper

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -457,4 +643,26 @@ private void workerLostHandler(long id) {
+  private void setMasterIdAndHostname() {
+    mMasterHostname = Configuration.get(PropertyKey.MASTER_HOSTNAME);
+    try {
+      RetryUtils.retry("get master id",
```
Is there somewhere where this master logs its own master id?

LuQQiu replied:
I am not quite sure why do the master need to log its id?

The id is just a representative of a master and it is meaningless to clients or programmers. Each time we will try to get an id for a standby master and if fails, we will assign a new id and ask the master to register with configuration. 

In addition, I do not find that we log the worker id.

Is there a need for logging the master id? Thanks!

gpang replied:
It would be helpful for debugging anything.

I think we log the worker id on the worker: https://github.com/Alluxio/alluxio/blob/v1.7.1/core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java#L232

LuQQiu replied:
In AlluxioMasterProcess.startMasters, when this master is a standby master, it will log:
```
LOG.info("Standby master with id {} starts sending heartbeat to leader master.", mMasterId);
```
Thanks!

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -165,7 +232,13 @@
+      mExecutorService = ExecutorServiceFactories
+          .fixedThreadPoolExecutorServiceFactory(Constants.META_MASTER_NAME, 2).create();
+      mMetaMasterClient = new MetaMasterMasterClient(MasterClientConfig.defaults());
+      setMasterIdAndHostname();
```
Does the primary heartbeat to itself?

LuQQiu replied:
In the AlluxioMasterProcess.startMasters(), if a master is leader master, it will setup lostMasterDectationHeartbeatExecutor to detect lost standby masters. If a master is not a leader master, it will setup its MetaMasterSync to getMasterId, register and heartbeat with the leader master.
``` java
    try {
      mWorkerConfigChecker.reset();
      if (isLeader) {
        mSafeModeManager.notifyPrimaryMasterStarted();

        //  The service that detects lost standby master nodes
        mExecutorService.submit(new HeartbeatThread(
            HeartbeatContext.MASTER_LOST_MASTER_DETECTION,
            new LostMasterDetectionHeartbeatExecutor(),
            (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      } else {
        // Standby master should setup MetaMasterSync to communicate with the leader master
        mMetaMasterSync = new MetaMasterSync(mMasterId, mMasterHostname, mMetaMasterClient);
        mExecutorService.submit(new HeartbeatThread(HeartbeatContext.MASTER_SYNC, mMetaMasterSync,
            (int) Configuration.getMs(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
      }
```
So primary will not send heartbeat/getMasterId/register to itself

LuQQiu commented:
@aaudiber @gpang I will use a separate PR to split AlluxioMasterProcess and DefaultMetaMaster after this PR. 

Main changes:
1. Changed MasterInfo to contains only id, hostname and lastUpdatedTimeMs.
2. Changed the masterHeart to return a command(Nothing, Register, Unknow) for extensibility.
3. Add shut down executor service when a master changes its position(leader <-> standby)

PTAL, thanks for the review!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19506/<h2>Build result: ABORTED</span></h2>[...truncated 4399 lines...][INFO] Alluxio Integration ................................ SUCCESS [  1.744 s][INFO] Alluxio Integration - Mesos ........................ SUCCESS [  7.200 s][INFO] Alluxio Assembly - Server .......................... SUCCESS [  5.753 s][INFO] Alluxio Integration Checker ........................ SUCCESS [ 27.656 s][INFO] Alluxio Core - Client - HDFS ....................... SUCCESS [ 12.707 s][INFO] Alluxio Core - Client - Runtime .................... SUCCESS [ 11.191 s][INFO] Alluxio Integration - FUSE ......................... SUCCESS [ 11.147 s][INFO] Alluxio MiniCluster ................................ SUCCESS [12:15 min][INFO] Alluxio Tests ...................................... FAILURE [  01:39 h][INFO] Alluxio Under File System - GCS .................... SKIPPED[INFO] Alluxio Under File System - HDFS ................... SKIPPED[INFO] Alluxio Under File System - Aliyun OSS ............. SKIPPED[INFO] Alluxio Under File System - S3A .................... SKIPPED[INFO] Alluxio Under File System - Swift .................. SKIPPED[INFO] Alluxio Under File System - Microsoft Azure Blob Storage SKIPPED[INFO] Alluxio Integration - YARN ......................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 01:59 h[INFO] Finished at: 2018-05-07T15:36:32-07:00Archiving artifacts[INFO] Final Memory: 143M/2810M[INFO] ------------------------------------------------------------------------
Test FAILed.

gpang commented on core/common/src/main/java/alluxio/PropertyKey.java:
```diff
@@ -923,11 +923,20 @@ public String toString() {
       new Builder(Name.MASTER_HEARTBEAT_INTERVAL_MS)
           .setAlias(new String[]{"alluxio.master.heartbeat.interval.ms"})
-          .setDefaultValue("1sec")
+          .setDefaultValue("1min")
```
Why is this so long? I would think something like 5 or 10 seconds should work?

LuQQiu replied:
Change to 10 seconds. We only need the heartbeat for config checker, so we may not need the frequent heartbeat.

gpang commented on core/common/src/main/java/alluxio/PropertyKey.java:
```diff
@@ -923,11 +923,20 @@ public String toString() {
+  public static final PropertyKey MASTER_HEARTBEAT_TIMEOUT_MS =
+      new Builder(Name.MASTER_HEARTBEAT_TIMEOUT_MS)
+          .setAlias(new String[]{"alluxio.master.heartbeat.timeout.ms"})
+          .setDefaultValue("15min")
```
Why is this so long? Since we just use the master heartbeat to monitor liveness, I think it should be fine to make this shorter, like 1 minute.

LuQQiu replied:
Change to 1 min, thanks

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -165,7 +239,13 @@
+      mExecutorServiceFactory = ExecutorServiceFactories
+          .fixedThreadPoolExecutorServiceFactory(Constants.META_MASTER_NAME, 2);
+      mMetaMasterClient = new MetaMasterMasterClient(MasterClientConfig.defaults());
+      setMasterIdAndHostname();
```
This looks like this is called whether or not this master is leader or standby. Does this mean the leader must connect to itself to get a master id? Will this work?

LuQQiu replied:
I guess the failed test may be because of this issue. I will change this part. When a master is a standby master, it will set its id.

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -232,6 +312,73 @@ public boolean isServing() {
+      return MetaCommand.Register;
+    }
+
+    synchronized (master) {
```
is this `synchronized` required?

LuQQiu replied:
Deleted, thanks

LuQQiu commented on core/common/src/main/java/alluxio/exception/NoMasterException.java:
```diff
@@ -0,0 +1,63 @@
+ * The exception thrown when a master does not exist, i.e. the id of the master cannot be found.
+ */
+@ThreadSafe
+public class NoMasterException extends AlluxioException {
```
MetaMasterMasterServiceHandler
```
  public RegisterMasterTResponse registerMaster(final long masterId,
      RegisterMasterTOptions options) throws AlluxioTException {
    return RpcUtils.call(LOG, new RpcUtils.RpcCallable<RegisterMasterTResponse>() {
      @Override
      public RegisterMasterTResponse call() throws AlluxioException {
        mMasterProcess.masterRegister(masterId, options);
        return new RegisterMasterTResponse();
      }
```
Because of the ServiceHandler, I can only throw AlluxioException here. NoMasterException and NoWorkerException(used by workerRegister) is under alluxio.exception package, so I can use them here.  Those exceptions under alluxio.exception.status packages cannot be throw here.

aaudiber replied:
You can use `RpcCallableThrowsIOException`, see https://github.com/Alluxio/alluxio/blob/master/core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java#L138

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19523/
Test PASSed.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,146 @@
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.concurrent.NotThreadSafe;
```
Changed, thanks!

LuQQiu commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  /**
+   * Handles a leader master command. The command is one of Unknown, Nothing, Register.
```
Delete the list of types of commands to make it easy to maintain.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19525/
Test PASSed.

LuQQiu commented:
@aaudiber @gpang  PTAL, thanks!
Main changes:
1. fix the bugs that leader master use thrift to getId from leader master
2. log master id
3. add stopExecutorService when a master gain/lost leadership and in stopMasters()

Things still need to discuss:

Should I use the masterHostname to represent a master or use bindHost instead?
masterHostname is normally used by clients, but In HA mode masters need their own alluxio.master.hostname so that they can register themselves with zookeeper.

If masterHostname always exists and no two masters use the same masterHostname, using masterHostname may not cause big issues here.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19526/
Test PASSed.

aaudiber commented:
Masters may have the same hostname in test environments. We need to do something similar to what we do for workers where we include the port as a part of the address. 

Bind host is an address used internally by the master when it decides what network interface to listen on.
Connect host (e.g. alluxio.master.hostname) is an externally accessible address that the master should be reachable at. This should be used by clients, or anything else that wants to interact with the master.
Usually these are the same, but in some places such as EC2 they might need to be different. In EC2, every instance has a public IP and a private IP. The instance itself doesn't know about its public IP, but the AWS networking stack knows to redirect requests for the public IP to the right machine. So if you tell an alluxio master to listen on (bind to) the public IP, it won't be able to. The master can only listen on the private IP for its machine. However, clients outside of EC2 need to use the public IP so that they can interact with the master.

LuQQiu commented:
@aaudiber Thanks for the detailed explanation! I am ready with this PR. 
In next PR, we create meta master and create one MasterAddress wire type that include alluxio.master.hostname and master RPC port?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19527/
Test PASSed.

apc999 commented on core/common/src/main/java/alluxio/PropertyKey.java:
```diff
@@ -923,11 +923,20 @@ public String toString() {
           .build();
+  public static final PropertyKey MASTER_HEARTBEAT_TIMEOUT_MS =
+      new Builder(Name.MASTER_HEARTBEAT_TIMEOUT_MS)
+          .setAlias(new String[]{"alluxio.master.heartbeat.timeout.ms"})
```
`setAlias` is for backward compatibility. For newly added properties, we don't need this.

apc999 commented on core/server/master/src/main/java/alluxio/master/MasterProcess.java:
```diff
@@ -92,4 +95,29 @@ private Factory() {} // prevent instantiation
+   * @param hostname the master hostname
+   * @return the master id for this master
+   */
+  long getMasterId(String hostname);
```
can you add a TODO somewhere in this file that we should have a separate `MetaMaster` interface and implementation, rather than free-riding `MasterProcess`

LuQQiu replied:
Added, thanks!

apc999 commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterMasterClient.java:
```diff
@@ -0,0 +1,108 @@
+ */
+@ThreadSafe
+public final class MetaMasterMasterClient extends AbstractMasterClient {
+  private MetaMasterMasterService.Client mClient = null;
```
nit: member variables are `null` by default, we don't set them to null typically.

LuQQiu replied:
We have to set it to null to avoid the mvn bugs:

MetaMasterMasterClient.mClient not initialized in constructor and dereferenced in alluxio.master.meta.MetaMasterMasterClient.lambda$getId$0(String) [alluxio.master.meta.MetaMasterMasterClient] At MetaMasterMasterClient.java:[line 78]

apc999 commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,147 @@
+   * @param masterClient the meta master client
+   */
+  public MetaMasterSync(AtomicReference<Long>  masterId,
+      String masterHostname, MetaMasterMasterClient masterClient) throws IOException {
```
typically, we will avoid throw exception in a Java constructor to avoid half-baked class instances. You can leave a TODO to change this into a factory method

LuQQiu replied:
Added, thanks!

apc999 commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,147 @@
+        LOG.error("Failed to receive leader master heartbeat command.", e);
+      } else {
+        LOG.error("Failed to receive or execute leader master heartbeat command: {}",
+            command.toString(), e);
```
nit: unnecessary `command.toString()` but `command` should be fine

LuQQiu replied:
Done, thanks!

apc999 commented on core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java:
```diff
@@ -0,0 +1,147 @@
+      if (command == null) {
+        LOG.error("Failed to receive leader master heartbeat command.", e);
+      } else {
+        LOG.error("Failed to receive or execute leader master heartbeat command: {}",
```
is it still possible "fail to receive" in this case?

LuQQiu replied:
Changed to "Failed to execute", thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19542/<h2>Build result: FAILURE</span></h2>[...truncated 2398 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19547/
Test PASSed.

LuQQiu commented:
@apc999 PTAL, thanks!
